File: src/test/scala/rps/GameTest.scala
package rps.test

import rps.Game.evaluatePlay
import rps.model.{Move, Result}
import Move._
import Result._

class GameTest extends munit.FunSuite {

  case class Play(userMove: Move, computerMove: Move)

  def check(
      name: String,
      inputs: List[Play],
      expected: Result
  )(using loc: munit.Location): Unit = {
    test(name) {
      inputs
        .map(play => evaluatePlay(play.userMove, play.computerMove))
        .foreach(assertEquals(_, expected))
    }
  }

  check(
    "player win",
    List(Play(Paper, Rock), Play(Rock, Scissors), Play(Scissors, Paper)),
    Win
  )

  check(
    "computer win",
    List(
      Play(Rock, Paper),
      Play(Scissors, Rock),
      Play(Paper, Scissors)
    ),
    Lose
  )

  check(
    "draw",
    List(Play(Paper, Paper), Play(Rock, Rock), Play(Scissors, Scissors)),
    Draw
  )
}

File: ts-rps/game.test.ts
import { expect, test } from "vitest";
import { playLogic } from "./game";
import { read } from "./model/Move";

test("game logic test", async () => {
  expect(await playLogic("Rock", "Paper")).toContain("You lose :<");
  expect(await playLogic("Rock", "Scissors")).toContain("You Win!!!");
  expect(await playLogic("Paper", "Rock")).toContain("You Win!!!");
  expect(await playLogic("Paper", "Scissors")).toContain("You lose :<");
  expect(await playLogic("Scissors", "Rock")).toContain("You lose :<");
  expect(await playLogic("Scissors", "Paper")).toContain("You Win!!!");
});

test("read function test", () => {
  expect(read("0")).toBe("Rock");
  expect(read("1")).toBe("Paper");
  expect(read("2")).toBe("Scissors");

  const randomNum = String(Math.round(Math.random()) + 3);
  expect(() => read(randomNum)).toThrowError();
});
